Add `selector-anb-no-unmatchable` rule

mattxwang opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes #5907.

> Is there anything in the PR that needs further explanation?

I've left a few comments in review. 

In addition, I'm a bit confused about [Grid-Structural Selectors](https://www.w3.org/TR/selectors-4/#table-pseudos). The spec itself doesn't have an example of `:nth-col()` or `:nth-last-col()`, the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-col)' code example doesn't render for me on either Chrome or Firefox and has no browser compat data, and I haven't found much online on correct use. I must be missing something - some guidance would be helpful!

mattxwang commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,117 @@
+					cssTreeSelector = parse(selector, { context: 'selector', positions: true });
+					checkSelector(cssTreeSelector);
+				} catch (e) {
+					report({
+						message: messages.rejectedParseError(ruleNode.selector),
+						node: ruleNode,
+						index: 0,
+						endIndex: 0,
+						result,
+						ruleName,
+					});
```
Wasn't exactly sure what to write here / what to write tests for. Are there other patterns that I should look at?

mattxwang commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,75 @@
+
+Disallow unmatchable An+B selectors.
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(0n+0) {}
+/*↑             ↑
+ * This unmatchable An+B selector */
+```
+
+[An+B selectors](https://drafts.csswg.org/css-syntax/#anb) are one-indexed. Selectors that always evaluate to `0` will not match any elements.
````
Any suggestions on a better thing to highlight / extended description?

Mouvedia replied:
question

Should it link to a draft?
Would https://www.w3.org/TR/css-syntax-3/#anb-microsyntax be better?

Mouvedia commented on docs/user-guide/rules.md:
```diff
@@ -73,6 +73,12 @@ Disallow overrides with these `no-overrides` rules.
+
+Disallow unmatchable things with these `no-unmatchable` rules.
+
+- [`selector-anb-no-unmatchable`](../../lib/rules/selector-anb-no-unmatchable/README.md): Disallow unmatchable An+B selectors
```
nit

full stop missing

Mouvedia commented on docs/user-guide/rules.md:
```diff
@@ -73,6 +73,12 @@ Disallow overrides with these `no-overrides` rules.
 
+### Unmatchable
+
+Disallow unmatchable things with these `no-unmatchable` rules.
```
minor

"things" should be replaced

mattxwang replied:
Gotcha, what's a good replacement? I'm not the best at writing docs 😅

Mouvedia replied:
Me neither. Wait for jeddy3.

jeddy3 replied:
Funnily enough, we've used "thing(s)" since the beginning of [our docs](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/rules.md#define-the-rule)

> the [thing](http://apps.workflower.fi/vocabs/css/en) the rule applies to, e.g. at-rule

We haven't been able to think of anything better. If anyone does, let's roll it out across all the instances of things in another PR. At the moment, let's keep it as things in this PR to be consistent with the other sections in the doc.

Mouvedia commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,117 @@
+
+const messages = ruleMessages(ruleName, {
+	rejected: (pseudoClass) => `Unexpected unmatchable An+B selector "${pseudoClass}"`,
+	rejectedParseError: (selector) => `Cannot parse selector "${selector}"`,
```
question

Iv seen one rejectedParseError test for the `declaration-property-value-no-unknown` rule.
Does it need to be covered?

Mouvedia commented on lib/rules/selector-anb-no-unmatchable/__tests__/index.js:
```diff
@@ -0,0 +1,145 @@
+		},
+	],
+
+	reject: [
```
question

What happens with https://www.w3.org/TR/css-syntax-3/#example-3d25855e?

mattxwang replied:
it errors; I'll add some test cases for this!

Mouvedia commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,117 @@
+							report({
+								message: messages.rejected(pseudoClassSelector.name),
+								node: ruleNode,
+								index: pseudoClassSelector.loc?.start.column,
```
question

Is `undefined` allowed as an index?

mattxwang replied:
I don't think so; there's a "soft" guarantee that if we include `positions` in the parse call, then `loc` is guaranteed to exist. Is there a better way for me to encode this? Ex, should I throw a parse error when `loc` doesn't exist?

ybiquitous replied:
Even if passing `undefined` to `index` and `endIndex`, I think there is no inconvenience. Because the lint problem location (`column` and `endColumn` will point to the whole selector. Also, it seems a rare case.

Mouvedia commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,117 @@
+				}
+			});
+
+			function checkSelector(/** @type {import('css-tree').CssNode} */ selector) {
```
minor

It's used at least twice; maybe @typedef it.

Mouvedia commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,117 @@
+
+			ruleNode.selectors.forEach((selector) => {
+				/** @type {import('css-tree').CssNode} */
+				let cssTreeSelector;
```
It should be a `const` at l58.

ybiquitous commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,121 @@
+				try {
+					const cssTreeSelector = parse(selector, { context: 'selector', positions: true });
+
+					checkSelector(cssTreeSelector);
```
[suggest] Moving `checkSelector(cssTreeSelector)` out of the `try...catch` block seems better to me because the current code may catch an unexpected error by `checkSelector()` .

So, I suggest a refactoring like this:

```js
let cssTreeSelector;

try {
	cssTreeSelector = parse(selector, { context: 'selector', positions: true });
} catch {
	// ...
	return;
}

checkSelector(cssTreeSelector);
```

ybiquitous commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,121 @@
+
+					checkSelector(cssTreeSelector);
+				} catch (e) {
+					const index = ruleNode.selectors
+						.slice(0, i)
+						.map((sel) => sel.length)
+						.reduce((c, a) => c + a, 0);
+					const endIndex = index + selector.length;
+
+					report({
+						message: messages.rejectedParseError(selector),
+						node: ruleNode,
+						index,
+						endIndex,
```
[suggest]

- We may simply use `word` instead of calculating `index` and `endIndex` because a parse error seems to rarely occur.
- How about using `messageArgs`? See #6589

```suggestion
					report({
						message: messages.rejectedParseError,
						messageArgs: [selector],
						node: ruleNode,
						word: selector,
```

ybiquitous commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,121 @@
+
+						if (isUnmatchableNth(child.nth)) {
+							report({
+								message: messages.rejected(pseudoClassSelector.name),
```
[suggest]

- How about prepending a colon (`:`) to a selector in a reported message? Because a pseudo-selector is often described with `:` like [`:nth-child()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child). If doing so, we may need to calculate `index - 1`.
- I suggest using `messageArgs` also here.

```suggestion
								message: messages.rejected,
								messageArgs: [`:${pseudoClassSelector.name}`],
```

ybiquitous reviewed (commented):
@mattxwang Thanks for the pull request, great work! 👏🏼 

I've left minor suggestions, so please check them.

mattxwang commented:
Thank you for the helpful feedback @ybiquitous! I need to get in the habit of adding `messageArgs` to new rules, and I agree that adding the `:` can make it easier for users to understand the message.

I think I've addressed all the comments, let me know what you think!

ybiquitous reviewed (approved):
@mattxwang Thanks, LGTM! 👍🏼 

I would suggest waiting for @jeddy3's review.

jeddy3 commented on .changeset/cuddly-olives-learn.md:
```diff
@@ -0,0 +1,5 @@
+"stylelint": minor
+---
+
+Added: `selector-anb-no-unmatchable`
```
```suggestion
Added: `selector-anb-no-unmatchable` rule
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(0n) {}
````
```suggestion
a:nth-last-child(0n) {}
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(0n+0) {}
````
```suggestion
a:nth-of-type(0n+0) {}
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(0 of a) {}
````
```suggestion
a:nth-last-of-type(0 of a) {}
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+```css
+a:nth-child(0 of a) {}
+```
+
+<!-- prettier-ignore -->
+```css
+a:nth-last-child(0) {}
+```
+
+<!-- prettier-ignore -->
+```css
+a:nth-of-type(0) {}
+```
+
+<!-- prettier-ignore -->
+```css
+a:nth-last-of-type(0) {}
+```
````
```suggestion
```

Let's keep the examples to the minimum to explain the rule's intent.

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(1n) {}
````
```suggestion
a:nth-last-child(1n) {}
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(1n+0) {}
````
```suggestion
a:nth-of-type(1n+0) {}
```

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/README.md:
````diff
@@ -0,0 +1,77 @@
+
+<!-- prettier-ignore -->
+```css
+a:nth-child(0n+1) {}
````
```suggestion
a:nth-last-of-type(1 of a) {}
```

We can mirror the accept examples in the reject ones.

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/index.js:
```diff
@@ -0,0 +1,120 @@
+				try {
+					cssTreeSelector = parse(selector, { context: 'selector', positions: true });
+				} catch (e) {
+					report({
+						message: messages.rejectedParseError,
+						messageArgs: [selector],
+						node: ruleNode,
+						word: selector,
+						result,
+						ruleName,
+					});
+
```
```suggestion
```

Let's silently catch parse errors, as it's out of the scope of this rule.

CSSTree is less forgiving than the PostCSS selector parser, and this will likely flag non-standard language extensions.

jeddy3 reviewed (changes requested):
@mattxwang Thanks. It's looking great.

I've requested one significant change and some minor nits.

mattxwang commented:
Thanks for the feedback @jeddy3; I think I've addressed all of the comments!

Two quick questions:

1. should I add test cases for nonstandard selectors / situations where CSSTree will swallow the error? basically, the rule would accept/bail out here
    - as a note, this will increase the test coverage!
3. should I also be doing grid-structural selectors? I was having some trouble finding good examples -- see PR description

jeddy3 commented:
Thanks for making the changes:

> should I add test cases for nonstandard selectors / situations where CSSTree will swallow the error?

Yeap, a single test case for that would be swell.

> should I also be doing grid-structural selectors? 

Let's leave them out for now.

jeddy3 commented on lib/rules/selector-anb-no-unmatchable/__tests__/index.js:
```diff
@@ -40,6 +40,14 @@ testRule({
+			code: 'a:nth-child(3n + - 6) {}',
+			description: 'CSSTree parse error not flagged',
+		},
+		{
+			code: '.mixin(@variable: x):nth-child(0) {}',
+			description: 'nonstandard syntax (LESS parametric mixin)',
+		},
```
```suggestion
```

Let's avoid adding CSS-like code in our tests.

mattxwang replied:
Ah! I think I misinterpreted your comment earlier, thanks for pointing this out. Have removed it.

jeddy3 reviewed (changes requested):
Thanks for making the changes. One minor request, then I think it LGTM.

jeddy3 reviewed (approved):
Thank you for making the changes.

LGTM.
